pkg/controller/vitesscell/reconcile_topo.go

```diff
@@ -36,64 +36,48 @@ const (
 	topoRequeueDelay = 5 * time.Second
 )
 
-func (r *ReconcileVitessCell) reconcileTopology(ctx context.Context, vtc *planetscalev2.VitessCell, keyspaces []*planetscalev2.VitessKeyspace) (topoEntries []string, err error) {
+func (r *ReconcileVitessCell) reconcileTopology(ctx context.Context, vtc *planetscalev2.VitessCell, ts *toposerver.Conn, keyspaces []*planetscalev2.VitessKeyspace) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
 	// Don't hold our slot in the reconcile work queue for too long.
 	ctx, cancel := context.WithTimeout(ctx, topoReconcileTimeout)
 	defer cancel()
 
-	if vtc.Status.Lockserver.Etcd != nil {
-		// We know something about local etcd status.
-		// We can use that to avoid trying to connect when we know it won't work.
-		if vtc.Status.Lockserver.Etcd.Available != corev1.ConditionTrue {
-			r.recorder.Event(vtc, corev1.EventTypeNormal, "TopoWaiting", "waiting for local etcd to become Available")
-			// Return success. We don't need to requeue because we'll get queued any time the EtcdCluster status changes.
-			return nil, nil
-		}
+	if *vtc.Spec.TopologyReconciliation.PruneSrvKeyspaces {
+		result, err := r.pruneSrvKeyspaces(ctx, vtc, keyspaces, ts)
+		resultBuilder.Merge(result, err)
 	}
 
-	// We actually know the address of the local lockserver already,
-	// but for now we'll follow the same rule as all Vitess components,
-	// which is to use the global lockserver to find the local ones.
-	ts, err := toposerver.Open(ctx, vtc.Spec.GlobalLockserver)
-	if err != nil {
-		r.recorder.Eventf(vtc, corev1.EventTypeWarning, "TopoConnectFailed", "failed to connect to global lockserver: %v", err)
-		return nil, err
-	}
-	defer ts.Close()
+	return resultBuilder.Result()
+}
+
+func (r *ReconcileVitessCell) pruneSrvKeyspaces(ctx context.Context, vtc *planetscalev2.VitessCell, keyspaces []*planetscalev2.VitessKeyspace, ts *toposerver.Conn) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
 
 	// Get the list of keyspaces deployed (served) in this cell.
 	srvKeyspaceNames, err := ts.GetSrvKeyspaceNames(ctx, vtc.Spec.Name)
 	if err != nil {
 		r.recorder.Eventf(vtc, corev1.EventTypeWarning, "TopoListFailed", "failed to list keyspaces in cell-local lockserver: %v", err)
-		return nil, err
+		return resultBuilder.Requeue()
```

Comment:
I meant `resultBuilder.RequeueAfter(topoRequeueDelay)` for all these. Sorry for not being clear. If we do just `Requeue()` it means "try again immediately", which will cause hot-looping whenever topo is down.

Reply:
Done https://github.com/planetscale/vitess-operator/pull/26/commits/af0eb34ebb84a1973e1cd5daec2a21eb3dc32cc5.